core/templates/services/html-length.service.ts

```diff
@@ -27,74 +27,102 @@ export class HtmlLengthService {
   constructor(
     private loggerService: LoggerService,
     private sanitizer: DomSanitizer
-  ) {}
+  ) { }
 
-  /**
-    * ['oppia-noninteractive-image', 'oppia-noninteractive-math']
-    * The above tags constitutes for non text nodes.
-    */
   nonTextTags = ['oppia-noninteractive-math', 'oppia-noninteractive-image'];
-
-  /**
-    * ['li','blockquote','em','strong', 'p', 'a']
-    * The above tags serve as the tags for text content. Blockquote,
-    * em, a, strong occur as descendants of p tag. li tags are
-    * descendants of ol/ul tags.
-    */
   textTags = ['p', 'ul', 'ol'];
 
   computeHtmlLengthInWords(htmlString: string): number {
     if (!htmlString) {
       this.loggerService.error('Empty string was passed to compute length');
       return 0;
     }
+
     const sanitizedHtml = this.sanitizer.sanitize(
       SecurityContext.HTML, htmlString) as string;
-    let totalWeight = this.calculateBaselineLength(sanitizedHtml);
+    let totalWords = this.calculateBaselineLength(sanitizedHtml);
 
-    // Identify custom tags using regex on the original HTML string.
-    const customTagRegex = /<oppia-noninteractive-(?:math|image)[^>]*>/g;
-    const customTags = htmlString.match(customTagRegex);
+    const customTags = this.extractCustomTags(htmlString);
+    for (const customTag of customTags) {
+      totalWords += this.getWeightForNonTextNodes(customTag, false);
+    }
 
-    if (customTags) {
-      for (const customTag of customTags) {
-        totalWeight += this.getWeightForNonTextNodes(customTag);
-      }
+    return totalWords;
+  }
+
+  computeHtmlLengthInCharacters(htmlString: string): number {
+    if (!htmlString) {
+      this.loggerService.error('Empty string was passed to compute length');
+      return 0;
     }
-    return totalWeight;
+
+    const sanitizedHtml = this.sanitizer.sanitize(
+      SecurityContext.HTML, htmlString) as string;
+    let totalCharacters = this.calculateBaselineLength(sanitizedHtml, true);
+
+    const customTags = this.extractCustomTags(htmlString);
+    for (const customTag of customTags) {
+      totalCharacters += this.getWeightForNonTextNodes(customTag, true);
+    }
+
+    return totalCharacters;
   }
 
-  private calculateBaselineLength(sanitizedHtml: string): number {
-    let domparser = new DOMParser();
-    let dom = domparser.parseFromString(sanitizedHtml, 'text/html');
+  private calculateBaselineLength(
+      sanitizedHtml: string, countCharacters: boolean = false): number {
```

Comment:
I mentioned this before but please don't use a boolean here. https://github.com/oppia/oppia/issues/18505#issuecomment-1833047826

I'm going to stop reviewing until you have addressed all the guidance from my earlier comment on the issue and confirmed that you have done so.